### ISSUE TITLE
Paradigm displays duplicate word forms in single cell

### ISSUE THREAD

**Issue by eddieantonio** [UAlbertaALTLab/itwewina#39]
For example, for [wâpamêw](http://altlab.ualberta.ca/itwewina/detail/crk/eng/w%C3%A2pam%C3%AAw.html?no_compounds=true&lemma_match=true&e_node=-4894728097406222093), the Present Tense 1s → 2s, 1s → 2p, 1p → 2s, and 1p → 2p each display the same cooresponding wordform *twice*:

Speech Act Participants | Independent | Conjunct
-- | -- | --
1s → 2s | kiwâpamitin kiwâpamitin | ê-wâpamitân
1s → 2p | kiwâpamitinâwâw kiwâpamitinâwâw | ê-wâpamitakok
1p → 2s | kiwâpamitinân kiwâpamitinân | ê-wâpamitâhk
1p → 2p | kiwâpamitinân kiwâpamitinân | ê-wâpamitâhk

@aarppe, this could be an FST bug, but it can be fixed in NDS: basically, when more than one wordform matches a paradigm table, we can create a unique `set()` of the wordforms, then output all of them in order from shortest to longest.

**Comment by aarppe** [comment]
Based on the analysis below, this does not appear to be an FST bug, but likely duplicate cells in the *.layout files or duplicate lines in the *.paradigm file, which the layout cell specifications match.

But in principle, this could be an FST bug, which we have earlier seen to result from a missing restricting flag diacritic along the path through the LEXC files (typically with a @U.X.X@ flag, that is alone and thus not constrained by an appropriate flag earlier, and you often notice it only when generating - the analyses might be entirely ok).

However, for the independent forms in question, this does not seem to be the case, as the generator only creates one form per analysis (though one form, 'kiwâpamitinân', is ambiguous):

hfst-lookup -q src/analyser-gt-norm.hfst
kiwâpamitin
kiwâpamitin	wâpamêw+V+TA+Ind+Prs+1Sg+2SgO	0.000000

arppe$ hfst-lookup -q src/generator-gt-norm.hfst
wâpamêw+V+TA+Ind+Prs+1Sg+2SgO
wâpamêw+V+TA+Ind+Prs+1Sg+2SgO	kiwâpamitin	0.000000

hfst-lookup -q src/analyser-gt-norm.hfst
kiwâpamitinâwâw
kiwâpamitinâwâw	wâpamêw+V+TA+Ind+Prs+1Sg+2PlO	0.000000

hfst-lookup -q src/generator-gt-norm.hfst
wâpamêw+V+TA+Ind+Prs+1Sg+2PlO
wâpamêw+V+TA+Ind+Prs+1Sg+2PlO	kiwâpamitinâwâw	0.000000

hfst-lookup -q src/analyser-gt-norm.hfst
kiwâpamitinân
kiwâpamitinân	wâpamêw+V+TA+Ind+Prs+1Pl+2PlO	0.000000
kiwâpamitinân	wâpamêw+V+TA+Ind+Prs+1Pl+2SgO	0.000000

hfst-lookup -q src/generator-gt-norm.hfst
wâpamêw+V+TA+Ind+Prs+1Pl+2PlO
wâpamêw+V+TA+Ind+Prs+1Pl+2PlO	kiwâpamitinân	0.000000

wâpamêw+V+TA+Ind+Prs+1Pl+2SgO
wâpamêw+V+TA+Ind+Prs+1Pl+2SgO	kiwâpamitinân	0.000000

hfst-lookup -q src/analyser-gt-norm.hfst
ê-wâpamitân
ê-wâpamitân	PV/e+wâpamêw+V+TA+Cnj+Prs+1Sg+2SgO	0.000000

hfst-lookup -q src/generator-gt-norm.hfst
PV/e+wâpamêw+V+TA+Cnj+Prs+1Sg+2SgO
PV/e+wâpamêw+V+TA+Cnj+Prs+1Sg+2SgO	ê-wâpamitân	0.000000

hfst-lookup -q src/analyser-gt-norm.hfst
ê-wâpamitakok
ê-wâpamitakok	PV/e+wâpamêw+V+TA+Cnj+Prs+1Sg+2PlO	0.000000

hfst-lookup -q src/analyser-gt-norm.hfst
ê-wâpamitâhk
ê-wâpamitâhk	PV/e+wâpamêw+V+TA+Cnj+Prs+1Pl+2PlO	0.000000
ê-wâpamitâhk	PV/e+wâpamêw+V+TA+Cnj+Prs+1Pl+2SgO	0.000000

hfst-lookup -q src/generator-gt-norm.hfst
PV/e+wâpamêw+V+TA+Cnj+Prs+1Pl+2PlO
PV/e+wâpamêw+V+TA+Cnj+Prs+1Pl+2PlO	ê-wâpamitâhk	0.000000

PV/e+wâpamêw+V+TA+Cnj+Prs+1Pl+2SgO
PV/e+wâpamêw+V+TA+Cnj+Prs+1Pl+2SgO	ê-wâpamitâhk	0.000000

There may be analyses for which one genuinely should get two forms generated from the FST, but a well-designed FST should not provide exactly the same string twice for the same unique analysis. Then, the linguistically correct option would be to present all unique strings (though a well-functioning FST should't require unique'ing), and as suggested from the shortest to the longest). Arok might suggest that we only show the shortest form in such a case, but let us leave that for later.

This was an issue earlier with NDS, when we allowed two alternative forms for the ê-conjuncts with vowel-initial stems, one with hyphens and the other with epenthesized, which Arok did not like at all. Then, we had Ryan hack this so that only hyphenated forms were output.

However, this has been fixed in the new crk FST, where epenthesized conjunct forms are considered substandard (receiving in descriptive analysis the +Err/Orth tag), which means that they are not accepted not generated by the normative analysers or generators.

**Comment by aarppe** [comment]
Ok - this is a *.paradigm file issue, cf.

sort verb-ta.paradigm | uniq -c | gawk '$1!=1'
...
   2 {{ lemma }}+V+TA+Ind+Prs+1Pl+2PlO
   2 {{ lemma }}+V+TA+Ind+Prs+1Pl+2SgO
   2 {{ lemma }}+V+TA+Ind+Prs+1Sg+2PlO
   2 {{ lemma }}+V+TA+Ind+Prs+1Sg+2SgO

Apparently, if the same form is specified for generation twice in the *.paradigm file, this are not uniqued by NDS, and then cells in the *.layout file are able to match more than one string/analysis.

Maybe we might want to have a check for duplicates in the *.paradigm files?

**Comment by eddieantonio** [comment]
Okay, so I'll delete some duplicates in the current version in giella, copy it over to itwêwina, and I'll create a quick duplicate checker within the giella infrastructure. I can also do a quick duplicate checker in Giella as well.

See: #40

**Comment by aarppe** [comment]
Perhaps there might be some benefit here from redundancy.

I'd think that the linguist ought to run the testing script for duplicates every time they edit a *.paradigm or *.layout file (as sometimes it could be intentional at least in the *.layout file) - but at the same time, we would want to sanity-check these on the itwêwina/NDS side just in case, if the linguist has accidentally missed something.  I didn't incorporate those code snippets into a shell-script yet, but the essential code is already there.

But generally, in the first instance it should be the linguist who removes or adds things in the *.paradigm and *.layout files.

**Comment by eddieantonio** [comment]
Okay, I added the consistency check on the giella side, and I'll add some checking on NDS side. I also corrected the verb-ta.paradigm, which simply had a "paragraph" erroneously duplicated.

**Comment by aarppe** [comment]
Okay as well. Added another script to evaluate how much matches a cell in a *.layout file has in the corresponding *.paradigm file.

./verify-paradigm-layout.sh verb-ta.paradigm verb-ta-full.layout
PARADIGM FILE: verb-ta.paradigm <- LAYOUT FILE: verb-ta-full.layout
Ind+Prs+1Sg+2SgO - 2 match(es): 23:{{ lemma }}+V+TA+Ind+Prs+1Sg+2SgO 28:{{ lemma }}+V+TA+Ind+Prs+1Sg+2SgO 
Ind+Prs+1Sg+2PlO - 2 match(es): 26:{{ lemma }}+V+TA+Ind+Prs+1Sg+2PlO 31:{{ lemma }}+V+TA+Ind+Prs+1Sg+2PlO 
Ind+Prs+1Pl+2SgO - 2 match(es): 24:{{ lemma }}+V+TA+Ind+Prs+1Pl+2SgO 29:{{ lemma }}+V+TA+Ind+Prs+1Pl+2SgO 
Ind+Prs+1Pl+2PlO - 2 match(es): 25:{{ lemma }}+V+TA+Ind+Prs+1Pl+2PlO 30:{{ lemma }}+V+TA+Ind+Prs+1Pl+2PlO 
Fut+Cond+3Sg+5Sg/PlO - 0 match(es): 
Fut+Cond+3Pl+5Sg/PlO - 0 match(es): 
Imp+Imm+2Sg+4Sg/PlO - 0 match(es): 
Imp+Del+2Sg+4Sg/PlO - 0 match(es): 
Imp+Imm+2Pl+4Sg/PlO - 0 match(es): 
Imp+Del+2Pl+4Sg/PlO - 0 match(es):

If there 0 matches, then the *.paradigm file doesn't generate the form (e.g. the ta- and ka- Conjunct forms previously). If there are 2 or more matches, then we get the duplicate forms in a cell (which may not always be the same strings), as the *.layout file allows for underspecifying the tag sequences.

**Comment by aarppe** [comment]
What we'd want are following checks wrt paradigm generation:

1. What is referred to in a *.layout file cell matches ideally one line in the corresponding *.paradigm file. If there's duplicate matches, then the reference in the *.layout file is too underspecified, or the *.paradigm file has duplicate form specifications.

2. Every line that is specified in the *.paradigm can be generated by the normative FST (there can be more than 1 legitimate form in some cases).

Not all forms in the *.paradigm files are checked in the YAML files, and not all forms in the YAML files are necessarily included in the *.paradigm files. But being able to check the extent of mismatch is generally good to know, even if it isn't necessarily a sign of error (at least as how we've specified the YAML files currently to keep things manageable).

**Comment by aarppe** [comment]
Implemented and check-in shell-script for verification task 1 above, cf.

```
./verify-paradigm-layout.sh verb-ta.paradigm verb-ta-full.layout
PARADIGM FILE: verb-ta.paradigm <- LAYOUT FILE: verb-ta-full.layout
-----
Ind+Prs+1Sg+2SgO - 2 match(es): 23:{{ lemma }}+V+TA+Ind+Prs+1Sg+2SgO 28:{{ lemma }}+V+TA+Ind+Prs+1Sg+2SgO 
Ind+Prs+1Sg+2PlO - 2 match(es): 26:{{ lemma }}+V+TA+Ind+Prs+1Sg+2PlO 31:{{ lemma }}+V+TA+Ind+Prs+1Sg+2PlO 
Ind+Prs+1Pl+2SgO - 2 match(es): 24:{{ lemma }}+V+TA+Ind+Prs+1Pl+2SgO 29:{{ lemma }}+V+TA+Ind+Prs+1Pl+2SgO 
Ind+Prs+1Pl+2PlO - 2 match(es): 25:{{ lemma }}+V+TA+Ind+Prs+1Pl+2PlO 30:{{ lemma }}+V+TA+Ind+Prs+1Pl+2PlO 
Fut+Cond+3Sg+5Sg/PlO - 0 match(es): ???
Fut+Cond+3Pl+5Sg/PlO - 0 match(es): ???
Imp+Imm+2Sg+4Sg/PlO - 0 match(es): ???
Imp+Del+2Sg+4Sg/PlO - 0 match(es): ???
Imp+Imm+2Pl+4Sg/PlO - 0 match(es): ???
Imp+Del+2Pl+4Sg/PlO - 0 match(es): ???
-----

./verify-paradigm-layout.sh verb-ta.paradigm | head -20
PARADIGM FILE: verb-ta.paradigm <- LAYOUT FILE: verb-ta-extended.layout
-----
Ind+Prs+1Sg+2SgO - 2 match(es): 23:{{ lemma }}+V+TA+Ind+Prs+1Sg+2SgO 28:{{ lemma }}+V+TA+Ind+Prs+1Sg+2SgO 
Ind+Prs+1Sg+2PlO - 2 match(es): 26:{{ lemma }}+V+TA+Ind+Prs+1Sg+2PlO 31:{{ lemma }}+V+TA+Ind+Prs+1Sg+2PlO 
Ind+Prs+1Pl+2SgO - 2 match(es): 24:{{ lemma }}+V+TA+Ind+Prs+1Pl+2SgO 29:{{ lemma }}+V+TA+Ind+Prs+1Pl+2SgO 
Ind+Prs+1Pl+2PlO - 2 match(es): 25:{{ lemma }}+V+TA+Ind+Prs+1Pl+2PlO 30:{{ lemma }}+V+TA+Ind+Prs+1Pl+2PlO 
FUTURE/INFINITIVE - 0 match(es): ???
-----

PARADIGM FILE: verb-ta.paradigm <- LAYOUT FILE: verb-ta-basic2.layout
-----
PRESENT TENSE - 0 match(es): ???
Speech Act Participant - 0 match(es): ???
2s → 1 - 0 match(es): ???
1s → 2 - 0 match(es): ???
Ind+Prs+1Sg+2SgO - 2 match(es): 23:{{ lemma }}+V+TA+Ind+Prs+1Sg+2SgO 28:{{ lemma }}+V+TA+Ind+Prs+1Sg+2SgO 
Mixed participant - 0 match(es): ???
1s → 3 - 0 match(es): ???
2s → 3 - 0 match(es): ???
3s → 1 - 0 match(es): ???
```

**Comment by eddieantonio** [comment]
I'm not sure I understand the output of this script.

**Edit**: if it gives output, that means something is not right, correct?

Also, when it finds a problem, it should exit with a non-zero status, so that `make check` can indicate the failure to the linguist. e.g.,

```
$ make check-paradigms
Duplicate entries in verb-ta.paradigm:
   2 {{ lemma }}+V+TA+Ind+Prs+1Sg+2PlO
make: *** [check-paradigms] Error 1
```

**Comment by aarppe** [comment]
Yes, the script outputs cells in *.layout which have either zero or two or more referents in the corresponding *.paradigm file.

I haven't implemented an exit status in the script output.

**Comment by aarppe** [comment]
Implemented exit status for the script 'verify-paradigm-layout.sh', as per my understanding of how that might work.

**Comment by eddieantonio** [comment]
Looks like this is resolved in the *.paradigm files!

To confirm, this is the output of running `make check-paradigm-uniqueness` in `$GTHOME/langs/crk/inc/paradigms`:

```
./ensure-unique-paradigm noun-in.paradigm
./ensure-unique-paradigm noun-na-d.paradigm
./ensure-unique-paradigm noun-na.paradigm
./ensure-unique-paradigm noun-ni-d.paradigm
./ensure-unique-paradigm noun-ni.paradigm
./ensure-unique-paradigm verb-ai.paradigm
./ensure-unique-paradigm verb-ta.paradigm
./ensure-unique-paradigm verb-ti.paradigm
Paradigms files OK!
```

**Comment by aarppe** [comment]
I actually used that verification script last night.

What needs to be yet checked is that the references in the *.layout files get unique matches in the *.paradigm files.

The layout verification script indicates that the short tag cell references (e.g. +Sg, +Pl) obviously get multiple matches (as many form have +Sg or +Pl), and so do some of the longer ones with a match both an undiminutivized and diminutive form.

My recollection of this was that Ryan had implemented a search option where one can specify that the tag sequence has to be matched in full (in contrast to partial matching), but whether that resolve this multiple matching would depend on the matching algorithm somehow knowing that the tag sequence must start after the lemma (which is not explictly specified as far as I understand in the *.paradigm and *.layout files, since we don't use pre-nouns preceding the lemma). 

My fuzzy recollection is that full feature sequence matching was specified by the '=' sign, and that there might have even been ways to refer to the beginning and end (whether with the '^' and '$' signs or something similar). A perfect example of why to document things...

This may, or may not be an issue, but I'm commenting this here as a historical reference.

**Comment by eddieantonio** [comment]
> My recollection of this was that Ryan had implemented a search option where one can specify that the tag sequence has to be matched in full (in contrast to partial matching), but whether that resolve this multiple matching would depend on the matching algorithm somehow knowing that the tag sequence must start after the lemma (which is not explictly specified as far as I understand in the *.paradigm and *.layout files, since we don't use pre-nouns preceding the lemma).

There is some code that appears to do this in [`paradigm_layouts.py`](https://github.com/UAlbertaALTLab/itwewina/blob/development/neahtta/paradigm_layouts.py#L195-L198). 

As far as the original issue is concerned, however, this has been fixed as of 88e90ccb.
